pkg/discovery/logsummary.go

```diff
@@ -0,0 +1,172 @@
+		logPatternNameErrors: LogPattern{
+			[](*regexp.Regexp){
+				regexp.MustCompilePOSIX(logFilePatternPodlogsString),
+			},
+			[](*regexp.Regexp){
+				regexp.MustCompilePOSIX(logPatternFailedString),
+				regexp.MustCompilePOSIX(logPatternErrorString),
+				regexp.MustCompilePOSIX(logPatternErrorCodeString),
+				regexp.MustCompilePOSIX(logPatternLevelErrorString),
+			},
+		},
+		logPatternNameWarnings: LogPattern{
+			[](*regexp.Regexp){
+				regexp.MustCompilePOSIX(logFilePatternPodlogsString),
+			},
+			[](*regexp.Regexp){
+				regexp.MustCompilePOSIX(logPatternWarningString),
+				regexp.MustCompilePOSIX(logPatternWarningCodeString),
+				regexp.MustCompilePOSIX(logPatternLevelWarningString),
+			},
+		},
+	}
+}
+
+func getPatternNamesForfile(relFilePath string, patterns LogPatterns) []string {
+	result := make([]string, 0)
+
+	for patternName, pattern := range patterns {
+		for _, fileNameRegex := range pattern.filePathPattern  {
+			if fileNameRegex.MatchString(relFilePath) {
+				result = append(result, patternName)
+			}
+		}
+	}
+	return result
+}
+
+
+// ReadLogSummary will recursively scan the tarballRootDir
+// looking for files with namess matching pre-defined regular expressions
+// and scanning the content of these files for pre-defined error matching regexes
+// and counting the number of hits
+// The return value is a map of results with key the type of condition
+// and the values of the map are the file names and the hit count
+// the patterns parameter is a list of LogPatterns objects that define what to scan for
+// The GetDefaultLogPatterns can be used to obtain such list.
+func ReadLogSummary(tarballRootDir string, patterns LogPatterns) (LogSummary, error) {
+	
+	logSummary := make(LogSummary)
+
+	findAndScanLogFiles := func (filePath string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if ! info.IsDir() {
+			relFilePath, err := filepath.Rel(tarballRootDir, filePath)
+			if err != nil {
+				logrus.Errorf("findAndScanLogFiles: ignoring file '%s' because getting its relative path to '%s' failed: %s", filePath, tarballRootDir, err)
+				return nil
+			}
+			patternsForFile := getPatternNamesForfile(relFilePath, patterns)
+			//If the patternsForFile list is empty it means this file is not interesting
+			if len(patternsForFile) == 0 {
+				return nil
+			}
+
+			file, err := os.Open(filePath)
+			if err != nil {
+				logrus.Errorf("findAndScanLogFiles: ignoring file '%s' because scanning it failed: %s", filePath, err)
+				return nil
+			}
+			defer file.Close()
+
+			scanner := bufio.NewScanner(file)
+			for scanner.Scan() {
+				line := scanner.Text()
+				for _, patternName := range patternsForFile {
+					for _, pattern := range patterns[patternName].matchPatterns {
+						if pattern.MatchString(line) {
+							if _, ok := logSummary[patternName]; !ok {
+								logSummary[patternName] = make(LogHitCounter)
+							}
+							logSummary[patternName][relFilePath]++
```

Comment:
If we match a pattern on a line, lets not keep increasing the count if it matches more patterns. I think we're double counting a fair number. Totally fine if a line hits as both a warning and an error but for instance:

```
W0211 17:21:27.983694       1 warnings.go:70] v1 ComponentStatus is deprecated in v1.19+
```
Is hitting 2x; once for W0211 and once for `warnings.go` hitting the `[wW]arn` regexp.